### PR TITLE
Optimize ICollection iteration (enumerator -> for)

### PR DESCRIPTION
**Pull request by isc30**
This is a POC of a simple compiler optimization.

`IEnumerable` is used to abstract a source of data of any kind (including generators). This forces the usage of the enumeration pattern for iterations, leaving us with a O(n) complexity for accessing data.

`ICollection` is quite similar, but it abstracts **in-memory** data structures (Array, List, etc). Knowing that the data is already in memory, we can use the `operator[]` to access data in a constant complexity O(1) (this really depends on the implementation, but applies to Arrays and Lists).

When a class implements `ICollection` or the resolved type from an Expression is `ICollection`, we can optimize every iteration (foreach) **from the enumerator pattern to a simple `for`** since we know that the data is already in memory represented as an Array in javascript.

The unit tests are self-explanatory 😄
Example of the optimization in [tests/TestCases/Statement/Foreach/Baseline.txt](https://github.com/DerivcoIpswich/dsharp/pull/112/files#diff-3910309817447598fd086dd6bc4f2579)

Waiting for any feedback about this

### CODE REVIEW

File: src/Core/Compiler/Extensions/TypeSymbolExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                    interfaces.AddRange(interfaceSymbol.Interfaces);
+                }
+
+                return interfaces;
+            }
+
+            return null;
+        }
+
+        internal static bool IsCollectionType(this TypeSymbol symbol)
```
Comment (fred-perkins) [inline src/Core/Compiler/Extensions/TypeSymbolExtensions.cs]: We could shorten this method with null chaining. 
```csharp
internal static bool IsCollectionType(this TypeSymbol symbol)
{
    return symbol?.GetInterfaces()
        ?.Any(i => i.FullName.StartsWith("System.Collections") && i.FullGeneratedName == "ICollection")
        ?? false;
}
```

File: src/Core/Compiler/Extensions/TypeSymbolExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace ScriptSharp.Extensions
+{
+    public static class TypeSymbolExtensions
+    {
+        internal static ICollection<InterfaceSymbol> GetInterfaces(this TypeSymbol symbol)
+        {
+            if (symbol is ClassSymbol classSymbol)
+            {
+                return classSymbol.Interfaces ?? new InterfaceSymbol[0];
```
Comment (fred-perkins) [inline src/Core/Compiler/Extensions/TypeSymbolExtensions.cs]: Can we use `Array.Empty<T>` here if possible.

File: src/Core/CoreLib/Array.cs [hunk 3]
```diff
@@ -17,14 +17,15 @@ namespace System {
...
+    public sealed class Array : ICollection {
 
         [ScriptField]
-        public int Length {
-            get {
-                return 0;
-            }
-        }
+        [ScriptName("length")]
+        public int Length => Count;
```
Comment (fred-perkins) [inline src/Core/CoreLib/Array.cs]: If we make these Extern we don't need to provide a body for the Get:
`public extern int Length { get; }`

File: src/Core/Compiler/Generator/StatementGenerator.cs [hunk 4]
```diff
@@ -125,6 +125,29 @@ internal static class StatementGenerator {
                 writer.Indent--;
                 writer.WriteLine("}");
             }
+            else if (statement.CollectionExpression.EvaluatedType.IsArray)
+            {
+                string dataSourceVariableName = statement.LoopVariable.GeneratedName;
+                string indexVariableName = dataSourceVariableName + "_index";
+
+                // var $$ = ({CollectionExpression});
+                writer.Write("var " + dataSourceVariableName + " = (");
```
Comment (fred-perkins) [inline src/Core/Compiler/Generator/StatementGenerator.cs]: Why do we open and close brackets around the collection we are using?

**Comment by fred-perkins** [comment]
Last little thing. Can you repoint these changes at the feature branch DSharp-SDK

File: src/Core/Compiler/Generator/StatementGenerator.cs (hunk 4 above)
Comment (isc30) [reply src/Core/Compiler/Generator/StatementGenerator.cs]: it's a good practice to inline expressions inside parenthesis to ensure that they don't affect the outer expression

File: src/Core/Compiler/Extensions/TypeSymbolExtensions.cs (hunk 1 above)
Comment (isc30) [reply src/Core/Compiler/Extensions/TypeSymbolExtensions.cs]: I find my version easier to maintain (add more logic) and debug, but up to you

Comment (fred-perkins) [reply src/Core/Compiler/Extensions/TypeSymbolExtensions.cs]: Honestly i don't mind at the end of the day :D

File: src/Core/Compiler/Generator/StatementGenerator.cs (hunk 4 above)
Comment (fred-perkins) [reply src/Core/Compiler/Generator/StatementGenerator.cs]: Oh really, learn something new every day. Where did you read up about that? Love to have a read myself :)

Comment (isc30) [reply src/Core/Compiler/Generator/StatementGenerator.cs]: I honestly have no idea where to find documentation about this.
I guess it's something you learn after fixing bugs caused by this in the past 😆

**Comment by isc30** [comment]
Closed in favour of https://github.com/DerivcoIpswich/dsharp/pull/116

**Pull request closed without merging**